### PR TITLE
Fix ImmutableConvention for records

### PR DESCRIPTION
**Pull request by TimHolzherr**
Using the `ImmutableConvtion` with records lead to a runtime error.
To make the `ImmutableConvention` work with records 3 things needed to be adjusted.
1. C# records add a hidden property [EqualityContract of type System.Type](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records) to each record instance.
The `ImmutableConvention` could not handle properties of type `System.Type`.
With this change properties of type `System.Type` are ignored.
2. The new init only setter needed special treatment because the underlying property is still marked with "CanWrite = true", even though the property needs to be treated as a read-only property.
3. Nullable parameter in the record initializer e.g. `public record Foo(string? nullableParameter)`, also need special treatment as the underlying property does not have the custom attribute `System.Runtime.CompilerServices.NullableAttribute`

Additionally I had trouble with tests interacting with each other.
After a lot of troubleshooting I found out that the RegisterConventionPack tests in `MongoDatabaseBuilderTests` cannot run at the same time as other tests which make use of ConventionPacks. I thus spited the Context.Test project into two projects, and moved the `MongoDatabaseBuilderTests` into the second project.

### CODE REVIEW

File: src/Context.Tests/ImmutableConventionTests.cs [hunk 1]
```diff
@@ -122,6 +122,23 @@ public async Task ApplyConvention_WithoutValueInDb_SerializeSuccessful()
                 result.MatchSnapshot();
             }
 
+            [Fact]
+            public async Task ApplyConvention_WithoutValueInDbWithoutDefault_SerializeSuccessful()
```
Comment (nscheibe) [inline src/Context.Tests/ImmutableConventionTests.cs:126]: I would create an additional test case, in which you InsertOneAsync with the C, like in ApplyConvention_WithoutValue_SerializeSuccessful, just to make sure in future, that the serializing works as well.

Comment (nscheibe) [reply src/Context.Tests/ImmutableConventionTests.cs:126]: Just a proposal

File: src/Context.Tests/ImmutableConventionWithRecordsTests.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            private readonly MongoDbContextData _context;
+
+            public SimpleRecordCase(MongoResource mongoResource)
+            {
+                _context = CreateContext(mongoResource);
+            }
+
+            [Fact]
+            public async Task ApplyConvention_SerializeSuccessful()
+            {
```
Comment (nscheibe) [inline src/Context.Tests/ImmutableConventionWithRecordsTests.cs:23]: I know the triple A is in the InsertAndFind method, but maybe we mark it with a comment // Arrange & Act & Assert

File: src/Context.Tests/ImmutableConventionWithRecordsTests.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            private readonly MongoDbContextData _context;
+
+            public RecordWithNullableCase(MongoResource mongoResource)
+            {
+                _context = CreateContext(mongoResource);
+            }
+
+            [Fact]
+            public async Task ApplyConvention_SerializeSuccessful()
+            {
```
Comment (nscheibe) [inline src/Context.Tests/ImmutableConventionWithRecordsTests.cs:42]: Eventually mark // Arrange & Act & Assert

File: src/Context.Tests/ImmutableConventionWithRecordsTests.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            private readonly MongoDbContextData _context;
+
+            public RecordWithNullableCase(MongoResource mongoResource)
+            {
+                _context = CreateContext(mongoResource);
+            }
+
+            [Fact]
+            public async Task ApplyConvention_SerializeSuccessful()
```
Comment (nscheibe) [inline src/Context.Tests/ImmutableConventionWithRecordsTests.cs:41]: Add maybe one test case where you fill both property (evt. a Theory)

File: src/Context.Tests/ImmutableConventionWithRecordsTests.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            [Fact]
+            public async Task ApplyConvention_WithoutValueInDb_SerializeSuccessful()
+            {
+                IMongoCollection<A> collectionTyped =
+                    _context.Database.GetCollection<A>("test");
+                IMongoCollection<BsonDocument> collectionUntyped =
+                    _context.Database.GetCollection<BsonDocument>("test");
+
+                // Act
+                await collectionUntyped.InsertOneAsync(new BsonDocument { { "Foo", "a" } });
```
Comment (nscheibe) [inline src/Context.Tests/ImmutableConventionWithRecordsTests.cs:55]: Maybe a second test, good case with both value filled. Just as good case... just a propsal.

**Review by nscheibe — APPROVED** [review]
Well Done! Some little comments or suggestions.

File: src/Context.Tests/ImmutableConventionTests.cs (hunk 1 above)
Comment (TimHolzherr) [reply src/Context.Tests/ImmutableConventionTests.cs:126]: I created `ApplyConvention_CtorWithDefault_SerializeSuccessful` to cover this case

File: src/Context.Tests/ImmutableConventionWithRecordsTests.cs (hunk 2 above)
Comment (TimHolzherr) [reply src/Context.Tests/ImmutableConventionWithRecordsTests.cs:23]: I added a comment on each usage of InsertAndFind

**Review by glucaci — APPROVED** [review]

**Review by nscheibe — COMMENTED** [review]
Well done

**Pull request merged**